vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/ExternalLinkManager.java

```diff
@@ -0,0 +1,124 @@
+import edu.asu.diging.vspace.core.model.impl.ExternalLink;
+import edu.asu.diging.vspace.core.model.impl.ExternalLinkValue;
+import edu.asu.diging.vspace.core.services.IExternalLinkManager;
+import edu.asu.diging.vspace.core.services.ISpaceManager;
+
+@Transactional
+@Service
+public class ExternalLinkManager extends LinkManager implements IExternalLinkManager{
+
+    @Autowired
+    private ISpaceManager spaceManager;
+
+    @Autowired
+    private ExternalLinkRepository externalLinkRepo;
+
+    @Autowired
+    private ExternalLinkDisplayRepository externalLinkDisplayRepo;
+
+    @Autowired
+    private IExternalLinkFactory externalLinkFactory;
+
+    @Autowired
+    private IExternalLinkDisplayFactory externalLinkDisplayFactory;
+
+    @Override
+    public List<ILinkDisplay> getLinkDisplays(String spaceId) {
+        return new ArrayList<>(externalLinkDisplayRepo.findExternalLinkDisplaysForSpace(spaceId));
+    }
+
+    @Override
+    protected ILink createLinkObject(String title, String id, IVSpaceElement target, String linkLabel) {
+        ISpace source = spaceManager.getSpace(id);
+        IExternalLink link = externalLinkFactory.createExternalLink(title, source, ((ExternalLinkValue)target).getValue());
+        link.setName(linkLabel);
+        externalLinkRepo.save((ExternalLink) link);
+        return link;
+    }
+
+    @Override
+    protected IVSpaceElement getTarget(String externalLink) {
+        return new ExternalLinkValue(externalLink);
+    }
+
+    @Override
+    protected ILinkDisplay updateLinkAndDisplay(ILink link, ILinkDisplay displayLink) {
+        externalLinkRepo.save((ExternalLink) link);
+        externalLinkDisplayRepo.save((ExternalLinkDisplay) displayLink);
+        return displayLink;
+    }
+
+    @Override
+    protected void setTarget(ILink link, IVSpaceElement target) {
+        link.setTarget(target);
+
+    }
+
+    @Override
+    protected ILinkDisplay getDisplayLink(String externalLinkDisplayId){
+        Optional<ExternalLinkDisplay> externalLinkOptional = externalLinkDisplayRepo.findById(externalLinkDisplayId);
+        if(!linksValidation(externalLinkOptional)) {
+            return null;
+        }
+        return externalLinkOptional.get();
+    }
+
+    @Override
+    protected ILink getLink(String externalLinkID){
+        Optional<ExternalLink> linkOptional = externalLinkRepo.findById(externalLinkID);
+        if(!linksValidation(linkOptional)) {
+            return null;
+        }
+        return linkOptional.get();
+    }
+
+    @Override
+    protected ILinkDisplay saveDisplayLinkRepo(ILink link, float positionX, float positionY, int rotation,
+            DisplayType displayType, byte[] linkImage, String imageFilename) throws ImageCouldNotBeStoredException {
+        ILinkDisplay displayLink = externalLinkDisplayFactory.createExternalLinkDisplay((IExternalLink)link);
+        setDisplayProperties(displayLink, positionX, positionY, rotation, displayType, linkImage, imageFilename);
```

Comment:
since that's the same for all links (it is, right?) it should be called in the superclass.

Reply:
Yes, but this needs to be called after the creation of DisplayLink and before saving link and DisplayLink. Also, to acieve this I would need to create two abstract methods in superclass. We did this to reduce number of methods in superclass.

Reply:
but this way you copy/paste the same code into all the subclasses just with slight modifications; that's more error prone that having more abstract methods that are pretty straight-forward to implement.